src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs

```diff
@@ -83,82 +78,28 @@ internal AsyncStreamReader(Process process, Stream stream, Action<string> callba
             _messageQueue = new Queue<string>();
         }
 
-        public void Close()
-        {
-            Dispose(true);
-        }
-
-        void IDisposable.Dispose()
-        {
-            Dispose(true);
-            GC.SuppressFinalize(this);
-        }
-
-        private void Dispose(bool disposing)
-        {
-            if (disposing)
-            {
-                if (_stream != null)
-                    _stream.Dispose();
-            }
-            if (_stream != null)
-            {
-                _stream = null;
-                _encoding = null;
-                _decoder = null;
-                _byteBuffer = null;
-                _charBuffer = null;
-            }
-
-            if (_eofEvent != null)
-            {
-                _eofEvent.Dispose();
-                _eofEvent = null;
-            }
-        }
-
-        public Encoding CurrentEncoding
-        {
-            get { return _encoding; }
-        }
-
-        public Stream BaseStream
-        {
-            get { return _stream; }
-        }
-
         // User calls BeginRead to start the asynchronous read
-        internal void BeginReadLine()
+        internal void BeginReadLine(CancellationToken token)
         {
-            if (_cancelOperation)
-            {
-                _cancelOperation = false;
-            }
-
             if (_sb == null)
             {
                 _sb = new StringBuilder(DefaultBufferSize);
-                _stream.ReadAsync(_byteBuffer, 0, _byteBuffer.Length).ContinueWith(ReadBuffer, CancellationToken.None, TaskContinuationOptions.None, TaskScheduler.Default);
+                ReadBuffer(token);
             }
             else
             {
-                FlushMessageQueue();
+                FlushMessageQueue(token);
             }
         }
 
-        internal void CancelOperation()
-        {
-            _cancelOperation = true;
-        }
-
         // This is the async callback function. Only one thread could/should call this.
-        private void ReadBuffer(Task<int> t)
+        private async void ReadBuffer(CancellationToken token)
```

Comment:
I think this function needs to main changes:
- The body should be wrapped in a Task.Run(async delegate { ... }).  That's because this call to ReadBuffer could be happening on the caller's thread, and without the ContinueWith forcing subsequent reads to run asynchronously, it's theoretically possible that this could become a long-running synchronous call.  With that, you also don't need the _eofEvent at all... you just store the scheduled Task, and the WaitForExit method just waits on the Task.
- The recursive call to ReadBuffer should be removed.  Instead, this function should just contain a loop: read, process, read, process, etc., and at some point either because 0 bytes were read or an exception occurred, the loop is exited and the ending code runs to flush the message queue.